File: 523.cpp
class Solution {
public:
    bool checkSubarraySum(vector<int>& nums, int k) {
        // 真是坑！
        
        k = k < 0 ? -k : k;
        unordered_map<int,int> hash;
        int sum = 0;
        for (int i = 0; i < nums.size(); i++) {
            sum += nums[i];
            if (k != 0) sum %= k;
            
            if (sum == 0 && i > 0) return true;
            else if (hash.count(sum) && i-hash[sum]>1) return true;
            else if (!hash.count(sum)) hash[sum] = i;
        }
        return false;
    }
};

File: 523.java
class Solution {
    public boolean checkSubarraySum(int[] nums, int k) {
        k = k < 0 ? -k : k;
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);
        int runningSum = 0;
        for (int i = 0; i < nums.length; i++) {
            runningSum += nums[i];
            if (k != 0) runningSum = runningSum % k;
            Integer prev = map.get(runningSum);
            if (prev != null) {
                if (i - prev > 1) return true;
            } else map.put(runningSum, i);
        }
        return false;
    }
}
